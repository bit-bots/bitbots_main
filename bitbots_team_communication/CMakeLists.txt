cmake_minimum_required(VERSION 3.5)
project(bitbots_team_communication)

find_package(ament_cmake REQUIRED)
find_package(ament_cmake_python REQUIRED)
find_package(rclpy REQUIRED)
find_package(std_msgs REQUIRED)
find_package(bitbots_msgs REQUIRED)
find_package(geometry_msgs REQUIRED)
find_package(tf2 REQUIRED)
find_package(tf2_ros REQUIRED)
find_package(tf2_geometry_msgs REQUIRED)
find_package(bitbots_utils REQUIRED)
find_package(bitbots_docs REQUIRED)

find_package(Protobuf REQUIRED)

protobuf_generate_python(
  PROTO_PY bitbots_team_communication/RobocupProtocol/robocup_extension.proto)

add_custom_target(
  bitbots_team_communication ALL
  DEPENDS ${PROTO_PY}
  COMMENT "Generating protobuf")
add_custom_command(
  TARGET bitbots_team_communication
  POST_BUILD
  COMMAND ${CMAKE_COMMAND} -E copy ${PROTO_PY}
          ${CMAKE_SOURCE_DIR}/bitbots_team_communication
  COMMENT "Copying protobuf to source dir")

enable_bitbots_docs()
ament_python_install_package(${PROJECT_NAME})

install(DIRECTORY launch DESTINATION share/${PROJECT_NAME})

install(DIRECTORY config DESTINATION share/${PROJECT_NAME})

install(DIRECTORY scripts/ USE_SOURCE_PERMISSIONS
        DESTINATION lib/${PROJECT_NAME})

if(BUILD_TESTING)
  find_package(ament_cmake_pytest REQUIRED)
  # Find all python test files
  file(GLOB_RECURSE pytest_tests_ test/**/test_*.py)
  # Add each test file to the ament test targets
  foreach(test_path_ ${pytest_tests_})
    get_filename_component(test_name_ ${test_path_} NAME_WE)
    ament_add_pytest_test(
      ${test_name_}
      ${test_path_}
      APPEND_ENV
      AMENT_PREFIX_INDEX=${ament_index_build_path}
      PYTHONPATH=${CMAKE_CURRENT_BINARY_DIR}
      TIMEOUT
      60
      WORKING_DIRECTORY
      "${CMAKE_SOURCE_DIR}")
  endforeach()
endif()

ament_package()
