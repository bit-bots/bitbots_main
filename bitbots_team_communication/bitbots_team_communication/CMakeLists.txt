cmake_minimum_required(VERSION 3.5)
project(bitbots_team_communication)

find_package(ament_cmake REQUIRED)
find_package(ament_cmake_python REQUIRED)
find_package(bitbots_docs REQUIRED)
find_package(Protobuf REQUIRED)

add_custom_target(${PROJECT_NAME}_generate_proto ALL)

add_custom_command(
  TARGET ${PROJECT_NAME}_generate_proto
  COMMAND protoc
  --pyi_out ${CMAKE_SOURCE_DIR}/bitbots_team_communication/
  --python_out ${CMAKE_SOURCE_DIR}/bitbots_team_communication/
  --proto_path ${CMAKE_SOURCE_DIR}/bitbots_team_communication/RobocupProtocol
  ${CMAKE_SOURCE_DIR}/bitbots_team_communication/RobocupProtocol/robocup_extension.proto
  COMMENT "Generating protobuf python interface including stubs")

enable_bitbots_docs()
ament_python_install_package(${PROJECT_NAME})

install(DIRECTORY launch DESTINATION share/${PROJECT_NAME})

install(DIRECTORY config DESTINATION share/${PROJECT_NAME})

install(DIRECTORY scripts/ USE_SOURCE_PERMISSIONS
        DESTINATION lib/${PROJECT_NAME})

if(BUILD_TESTING)
  find_package(ament_cmake_pytest REQUIRED)
  # Find all python test files
  file(GLOB_RECURSE pytest_tests_ test/**/test_*.py)
  # Add each test file to the ament test targets
  foreach(test_path_ ${pytest_tests_})
    get_filename_component(test_name_ ${test_path_} NAME_WE)
    ament_add_pytest_test(
      ${test_name_}
      ${test_path_}
      APPEND_ENV
      AMENT_PREFIX_INDEX=${ament_index_build_path}
      PYTHONPATH=${CMAKE_CURRENT_BINARY_DIR}
      TIMEOUT
      60
      WORKING_DIRECTORY
      "${CMAKE_SOURCE_DIR}")
  endforeach()

  find_package(ament_cmake_mypy REQUIRED)


  # The following logic for manually filtering out the files can be removed if 
  # https://github.com/ament/ament_lint/pull/516 is merged and released in our ROS distro
  
  # Collect all .py and .pyi files
  file(GLOB_RECURSE PY_FILES "${CMAKE_CURRENT_LIST_DIR}/**/*.py")
  file(GLOB_RECURSE PYI_FILES "${CMAKE_CURRENT_LIST_DIR}/**/*.pyi")

  # Filter out .py files if a corresponding .pyi file exists
  foreach(py_file IN LISTS PY_FILES)
      # Get the directory and basename (without extension)
      get_filename_component(py_dir "${py_file}" DIRECTORY)
      get_filename_component(py_basename "${py_file}" NAME_WE)

      # Construct the corresponding .pyi file path
      set(pyi_file "${py_dir}/${py_basename}.pyi")
      
      # If a corresponding .pyi file exists, remove the .py file from the list
      if(EXISTS "${pyi_file}")
          list(REMOVE_ITEM PY_FILES "${py_file}")
      endif()
  endforeach()

  ament_mypy(
    CONFIG_FILE "${CMAKE_CURRENT_LIST_DIR}/mypy.ini"
    ${PY_FILES} ${PYI_FILES}
    )
endif()

ament_package()
