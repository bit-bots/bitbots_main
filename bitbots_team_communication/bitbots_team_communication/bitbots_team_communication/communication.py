import socket
from ipaddress import IPv4Address
from typing import List, Optional

from rclpy.node import Node


class SocketCommunication:
    def __init__(self, node: Node, logger, team_id, robot_id):
        self.logger = logger

        self.buffer_size: int = 1024
        self.socket: Optional[socket.socket] = None
        self.target_ip: IPv4Address = IPv4Address(node.get_parameter("target_ip").value)

        if self.target_ip.is_loopback:
            # local mode on loopback device, bind to port depending on bot id and team id
            local_target_ports: List[int] = node.get_parameter("local_target_ports").value
            self.target_ports = [port + 10 * team_id for port in local_target_ports]
            self.receive_port = self.target_ports[robot_id - 1]
        else:
            target_port: int = node.get_parameter("target_port").value
            receive_port: int = node.get_parameter("receive_port").value
            self.target_ports = [target_port]
            self.receive_port = receive_port

    def __del__(self):
        self.close_connection()

    def is_setup(self) -> bool:
        return self.socket is not None

    def establish_connection(self):
        if not self.is_setup():
            self.socket = self.get_connection()

    def get_connection(self) -> socket.socket:
        self.logger.info(f"Binding to port {self.receive_port} to receive messages")
        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
        sock.bind(("0.0.0.0", self.receive_port))
        return sock

    def close_connection(self):
        if self.is_setup():
            self.socket.close()
            self.logger.info("Connection closed.")

    def receive_message(self) -> Optional[bytes]:
        self.assert_is_setup()
        msg, _, flags, _ = self.socket.recvmsg(self.buffer_size)
        is_message_truncated = flags & socket.MSG_TRUNC
        if is_message_truncated:
            self.handle_truncated_message()
            return None
        else:
            return msg

    def handle_truncated_message(self):
        self.logger.warn(
            f"recvmsg flag {socket.MSG_TRUNC} signaling a packet larger than buffer size {self.buffer_size}"
        )
        self.buffer_size *= 2
        self.logger.info(f"doubled buffer size to {self.buffer_size}")

    def send_message(self, message):
        self.assert_is_setup()
        for port in self.target_ports:
            self.logger.debug(f"Sending message to {port} on {self.target_ip}")
            self.socket.sendto(message, (str(self.target_ip), port))

    def assert_is_setup(self):
        assert self.is_setup(), "Socket is not yet initialized"
