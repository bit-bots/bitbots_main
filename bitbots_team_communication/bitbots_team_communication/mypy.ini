# Global options:

[mypy]
check_untyped_defs = True
ignore_missing_imports = True
exclude = .*_pb2\.py
