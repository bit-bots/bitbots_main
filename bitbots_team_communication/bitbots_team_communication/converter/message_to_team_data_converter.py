from typing import List, Tuple

import transforms3d
from geometry_msgs.msg import PoseWithCovariance
from numpy import double

import bitbots_team_communication.robocup_extension_pb2 as Proto  # noqa: N812
from bitbots_msgs.msg import RobotRelative, RobotRelativeArray, TeamData


class MessageToTeamDataConverter:
    def __init__(self, team_mapping, role_mapping, action_mapping, side_mapping):
        self.team_mapping = team_mapping
        self.role_mapping = role_mapping
        self.action_mapping = action_mapping
        self.side_mapping = side_mapping

    def convert(self, message: Proto.Message, team_data: TeamData) -> TeamData:
        team_data.robot_id = message.current_pose.player_id
        team_data.state = message.state

        team_data.robot_position = self.convert_robot_pose(message.current_pose)
        team_data.ball_absolute = self.convert_ball_pose(message.ball)

        team_data.robots = self.convert_robots(message.others, message.other_robot_confidence)
        team_data.robots.header = team_data.header

        return self.convert_optional_fields(message, team_data)

    def convert_optional_fields(self, message: Proto.Message, team_data: TeamData) -> TeamData:
        if hasattr(message, "time_to_ball"):
            team_data.time_to_position_at_ball = message.time_to_ball

        return self.convert_strategy(message, team_data)

    def convert_strategy(self, message: Proto.Message, team_data: TeamData) -> TeamData:
        if hasattr(message, "role"):
            team_data.strategy.role = self.role_mapping[message.role]
        if hasattr(message, "action"):
            team_data.strategy.action = self.action_mapping[message.action]
        if hasattr(message, "offensive_side"):
            team_data.strategy.offensive_side = self.side_mapping[message.offensive_side]

        return team_data

    def convert_robots(
        self, message_robots: List[Proto.Robot], message_robot_confidence: List[float]
    ) -> RobotRelativeArray:
        relative_robots = RobotRelativeArray()
        for index, robot in enumerate(message_robots):
            robot_relative = RobotRelative(player_number=robot.player_id, type=self.team_mapping[robot.team])
            robot_relative.pose.pose = self.convert_robot_pose(robot)

            if index < len(message_robot_confidence):
                robot_relative.pose.confidence = message_robot_confidence[index]

            relative_robots.robots.append(robot_relative)

        return relative_robots

    def convert_ball_pose(self, message_ball_pose: Proto.Ball) -> PoseWithCovariance:
        ball = PoseWithCovariance()
        ball.pose.position.x = message_ball_pose.position.x
        ball.pose.position.y = message_ball_pose.position.y
        ball.pose.position.z = message_ball_pose.position.z

        if message_ball_pose.covariance:
            self.convert_to_row_major_covariance(ball.covariance, message_ball_pose.covariance)

        return ball

    def convert_robot_pose(self, message_robot_pose: Proto.Robot) -> PoseWithCovariance:
        robot = PoseWithCovariance()
        robot.pose.position.x = message_robot_pose.position.x
        robot.pose.position.y = message_robot_pose.position.y

        quaternion = self.convert_to_quat((0, 0, message_robot_pose.position.z))
        robot.pose.orientation.w = quaternion[0]
        robot.pose.orientation.x = quaternion[1]
        robot.pose.orientation.y = quaternion[2]
        robot.pose.orientation.z = quaternion[3]

        self.convert_to_row_major_covariance(robot.covariance, message_robot_pose.covariance)

        return robot

    def convert_to_quat(self, euler_angles: Tuple[float, float, float]):
        return transforms3d.euler.euler2quat(*euler_angles)

    def convert_to_row_major_covariance(self, row_major_covariance: List[double], covariance_matrix: Proto.fmat3):
        # ROS covariance is row-major 36 x float, while protobuf covariance
        # is column-major 9 x float [x, y, Î¸]
        row_major_covariance[0] = covariance_matrix.x.x
        row_major_covariance[1] = covariance_matrix.y.x
        row_major_covariance[5] = covariance_matrix.z.x
        row_major_covariance[6] = covariance_matrix.x.y
        row_major_covariance[7] = covariance_matrix.y.y
        row_major_covariance[11] = covariance_matrix.z.y
        row_major_covariance[30] = covariance_matrix.x.z
        row_major_covariance[31] = covariance_matrix.y.z
        row_major_covariance[35] = covariance_matrix.z.z
