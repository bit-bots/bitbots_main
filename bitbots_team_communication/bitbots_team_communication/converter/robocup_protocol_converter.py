from enum import IntEnum

from soccer_vision_attribute_msgs.msg import Robot as RobotAttributes

import bitbots_team_communication.robocup_extension_pb2 as Proto  # noqa: N812
from bitbots_msgs.msg import RobotRelative, Strategy
from bitbots_team_communication.converter.message_to_team_data_converter import MessageToTeamDataConverter
from bitbots_team_communication.converter.state_to_message_converter import StateToMessageConverter


class TeamColor(IntEnum):
    BLUE = 0
    RED = 1


class RobocupProtocolConverter:
    def __init__(self, own_team_color: TeamColor):
        self.role_mapping = (
            (Proto.Role.ROLE_UNDEFINED, Strategy.ROLE_UNDEFINED),
            (Proto.Role.ROLE_IDLING, Strategy.ROLE_IDLING),
            (Proto.Role.ROLE_OTHER, Strategy.ROLE_OTHER),
            (Proto.Role.ROLE_STRIKER, Strategy.ROLE_STRIKER),
            (Proto.Role.ROLE_SUPPORTER, Strategy.ROLE_SUPPORTER),
            (Proto.Role.ROLE_DEFENDER, Strategy.ROLE_DEFENDER),
            (Proto.Role.ROLE_GOALIE, Strategy.ROLE_GOALIE),
        )
        self.action_mapping = (
            (Proto.Action.ACTION_UNDEFINED, Strategy.ACTION_UNDEFINED),
            (Proto.Action.ACTION_POSITIONING, Strategy.ACTION_POSITIONING),
            (Proto.Action.ACTION_GOING_TO_BALL, Strategy.ACTION_GOING_TO_BALL),
            (Proto.Action.ACTION_TRYING_TO_SCORE, Strategy.ACTION_TRYING_TO_SCORE),
            (Proto.Action.ACTION_WAITING, Strategy.ACTION_WAITING),
            (Proto.Action.ACTION_KICKING, Strategy.ACTION_KICKING),
            (Proto.Action.ACTION_SEARCHING, Strategy.ACTION_SEARCHING),
            (Proto.Action.ACTION_LOCALIZING, Strategy.ACTION_LOCALIZING),
        )
        self.side_mapping = (
            (Proto.OffensiveSide.SIDE_UNDEFINED, Strategy.SIDE_UNDEFINED),
            (Proto.OffensiveSide.SIDE_LEFT, Strategy.SIDE_LEFT),
            (Proto.OffensiveSide.SIDE_MIDDLE, Strategy.SIDE_MIDDLE),
            (Proto.OffensiveSide.SIDE_RIGHT, Strategy.SIDE_RIGHT),
        )

        self.proto_to_team_data_team_mapping = {
            Proto.Team.UNKNOWN_TEAM: RobotRelative.ROBOT_UNDEFINED,
            Proto.Team.BLUE: RobotRelative.ROBOT_BLUE,
            Proto.Team.RED: RobotRelative.ROBOT_RED,
        }
        self.state_to_proto_team_mapping = {
            RobotAttributes.TEAM_OWN: Proto.Team.RED if own_team_color == TeamColor.RED else Proto.Team.BLUE,
            RobotAttributes.TEAM_OPPONENT: Proto.Team.BLUE if own_team_color == TeamColor.RED else Proto.Team.RED,
            RobotAttributes.TEAM_UNKNOWN: Proto.Team.UNKNOWN_TEAM,
        }

        proto_to_team_data_mappings = {
            "team_mapping": self.proto_to_team_data_team_mapping,
            "role_mapping": dict(self.role_mapping),
            "action_mapping": dict(self.action_mapping),
            "side_mapping": dict(self.side_mapping),
        }

        def reverse_mapping(mapping):
            return dict((b, a) for a, b in mapping)

        state_to_proto_mappings = {
            "team_mapping": self.state_to_proto_team_mapping,
            "role_mapping": reverse_mapping(self.role_mapping),
            "action_mapping": reverse_mapping(self.action_mapping),
            "side_mapping": reverse_mapping(self.side_mapping),
        }

        self.convert_from_message = MessageToTeamDataConverter(**proto_to_team_data_mappings).convert
        self.convert_to_message = StateToMessageConverter(**state_to_proto_mappings).convert
