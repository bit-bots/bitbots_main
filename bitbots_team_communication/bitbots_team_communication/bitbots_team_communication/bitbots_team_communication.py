#!/usr/bin/env python3

import socket
import struct
import threading
from typing import List, Optional, Tuple

import rclpy
from ament_index_python.packages import get_package_share_directory
from bitbots_tf_buffer import Buffer
from bitbots_utils.utils import get_parameter_dict, get_parameters_from_other_node
from game_controller_hl_interfaces.msg import GameState
from geometry_msgs.msg import PoseWithCovarianceStamped, Twist, TwistWithCovarianceStamped
from numpy import double
from rclpy.callback_groups import MutuallyExclusiveCallbackGroup
from rclpy.duration import Duration
from rclpy.executors import MultiThreadedExecutor
from rclpy.node import Node
from rclpy.time import Time
from soccer_vision_3d_msgs.msg import Robot, RobotArray
from std_msgs.msg import Float32, Header
from tf2_geometry_msgs import PointStamped, PoseStamped
from tf2_ros import TransformException

import bitbots_team_communication.robocup_extension_pb2 as Proto  # noqa: N812
from bitbots_msgs.msg import Strategy, TeamData
from bitbots_team_communication.communication import SocketCommunication
from bitbots_team_communication.converter.robocup_protocol_converter import RobocupProtocolConverter, TeamColor


class TeamCommunication:
    def __init__(self):
        self._package_path = get_package_share_directory("bitbots_team_communication")
        self.node = Node("team_comm", automatically_declare_parameters_from_overrides=True)
        self.logger = self.node.get_logger()

        self.logger.info("Initializing bitbots_team_communication...")
        params_blackboard = get_parameters_from_other_node(
            self.node, "parameter_blackboard", ["bot_id", "team_id", "team_color"]
        )
        self.player_id = params_blackboard["bot_id"]
        self.team_id = params_blackboard["team_id"]
        self.team_color_id = params_blackboard["team_color"]

        self.protocol_converter = RobocupProtocolConverter(TeamColor(self.team_color_id))

        self.logger.info(f"Starting for {self.player_id} in team {self.team_id}...")
        self.socket_communication = SocketCommunication(self.node, self.logger, self.team_id, self.player_id)

        self.rate: int = self.node.get_parameter("rate").value
        self.lifetime: int = self.node.get_parameter("lifetime").value
        self.avg_walking_speed: float = self.node.get_parameter("avg_walking_speed").value

        self.topics = get_parameter_dict(self.node, "topics")
        self.map_frame: str = self.node.get_parameter("map_frame").value

        self.create_publishers()
        self.create_subscribers()

        self.set_state_defaults()

        self.tf_buffer = Buffer(self.node)

        self.run_spin_in_thread()
        self.try_to_establish_connection()

        self.node.create_timer(1 / self.rate, self.send_message, callback_group=MutuallyExclusiveCallbackGroup())
        self.receive_forever()

    def spin(self):
        multi_executor = MultiThreadedExecutor(num_threads=10)
        multi_executor.add_node(self.node)
        multi_executor.spin()

    def run_spin_in_thread(self):
        # Necessary in ROS2, else we are forever stuck receiving messages
        thread = threading.Thread(target=self.spin, daemon=True)
        thread.start()

    def set_state_defaults(self):
        self.gamestate: Optional[GameState] = None
        self.pose: Optional[PoseWithCovarianceStamped] = None
        self.cmd_vel: Optional[Twist] = None
        self.cmd_vel_time = Time(clock_type=self.node.get_clock().clock_type)
        self.ball: Optional[PointStamped] = None
        self.ball_velocity: Tuple[float, float, float] = (0, 0, 0)
        self.ball_covariance: List[double] = []
        self.strategy: Optional[Strategy] = None
        self.strategy_time = Time(clock_type=self.node.get_clock().clock_type)
        self.time_to_ball: Optional[float] = None
        self.time_to_ball_time = Time(clock_type=self.node.get_clock().clock_type)
        self.seen_robots: Optional[RobotArray] = None
        self.move_base_goal: Optional[PoseStamped] = None

    def try_to_establish_connection(self):
        # we will try multiple times till we manage to get a connection
        while rclpy.ok() and not self.socket_communication.is_setup():
            self.socket_communication.establish_connection()
            self.node.get_clock().sleep_for(Duration(seconds=1))

    def create_publishers(self):
        self.team_data_publisher = self.node.create_publisher(TeamData, self.topics["team_data_topic"], qos_profile=1)

    def create_subscribers(self):
        self.node.create_subscription(
            GameState,
            self.topics["gamestate_topic"],
            self.gamestate_cb,
            qos_profile=1,
            callback_group=MutuallyExclusiveCallbackGroup(),
        )
        self.node.create_subscription(
            PoseWithCovarianceStamped,
            self.topics["pose_topic"],
            self.pose_cb,
            qos_profile=1,
            callback_group=MutuallyExclusiveCallbackGroup(),
        )
        self.node.create_subscription(
            Twist,
            self.topics["cmd_vel_topic"],
            self.cmd_vel_cb,
            qos_profile=1,
            callback_group=MutuallyExclusiveCallbackGroup(),
        )
        self.node.create_subscription(
            PoseWithCovarianceStamped,
            self.topics["ball_topic"],
            self.ball_cb,
            qos_profile=1,
            callback_group=MutuallyExclusiveCallbackGroup(),
        )
        self.node.create_subscription(
            TwistWithCovarianceStamped,
            self.topics["ball_velocity_topic"],
            self.ball_velocity_cb,
            qos_profile=1,
            callback_group=MutuallyExclusiveCallbackGroup(),
        )
        self.node.create_subscription(
            Strategy,
            self.topics["strategy_topic"],
            self.strategy_cb,
            qos_profile=1,
            callback_group=MutuallyExclusiveCallbackGroup(),
        )
        self.node.create_subscription(
            Float32,
            self.topics["time_to_ball_topic"],
            self.time_to_ball_cb,
            qos_profile=1,
            callback_group=MutuallyExclusiveCallbackGroup(),
        )
        self.node.create_subscription(
            RobotArray,
            self.topics["robots_topic"],
            self.robots_cb,
            qos_profile=1,
            callback_group=MutuallyExclusiveCallbackGroup(),
        )
        self.node.create_subscription(
            PoseStamped,
            self.topics["move_base_goal_topic"],
            self.move_base_goal_cb,
            qos_profile=1,
            callback_group=MutuallyExclusiveCallbackGroup(),
        )

    def gamestate_cb(self, msg: GameState):
        self.gamestate = msg

    def pose_cb(self, msg: PoseWithCovarianceStamped):
        self.pose = msg

    def cmd_vel_cb(self, msg: Twist):
        self.cmd_vel = msg
        self.cmd_vel_time = self.get_current_time().to_msg()

    def strategy_cb(self, msg: Strategy):
        self.strategy = msg
        self.strategy_time = self.get_current_time().to_msg()

    def time_to_ball_cb(self, msg: Float32):
        self.time_to_ball = msg.data
        self.time_to_ball_time = self.get_current_time().to_msg()

    def move_base_goal_cb(self, msg: PoseStamped):
        self.move_base_goal = msg

    def robots_cb(self, msg: RobotArray):
        def transform_to_map(robot_relative: Robot):
            # @TODO: check if this is not handled by the transform itself
            robot_pose = PoseStamped(header=msg.header, pose=robot_relative.bb.center)
            try:
                robot_on_map = self.transform_to_map_frame(robot_pose)
                robot_relative.bb.center = robot_on_map.pose
                return robot_relative
            except TransformException as err:
                self.logger.error(f"Could not transform robot to map frame: {err}")

        robots_on_map = list(filter(None, map(transform_to_map, msg.robots)))
        self.seen_robots = RobotArray(header=msg.header, robots=robots_on_map)
        self.seen_robots.header.frame_id = self.map_frame

    def ball_cb(self, msg: PoseWithCovarianceStamped):
        ball_point = PointStamped(header=msg.header, point=msg.pose.pose.position)
        try:
            self.ball = self.transform_to_map_frame(ball_point)
            self.ball_covariance = msg.pose.covariance
        except TransformException as err:
            self.logger.error(f"Could not transform ball to map frame: {err}")

    def ball_velocity_cb(self, msg: TwistWithCovarianceStamped):
        self.ball_velocity = (
            msg.twist.twist.linear.x,
            msg.twist.twist.linear.y,
            msg.twist.twist.angular.z,
        )

    def transform_to_map_frame(self, field, timeout_in_s=0.3):
        return self.tf_buffer.transform(field, self.map_frame, timeout=Duration(seconds=timeout_in_s))

    def receive_forever(self):
        while rclpy.ok():
            try:
                message = self.socket_communication.receive_message()
            except (struct.error, socket.timeout):
                continue

            if message:
                self.handle_message(message)

    def handle_message(self, string_message: bytes):
        message = Proto.Message()
        message.ParseFromString(string_message)

        if self.should_message_be_discarded(message):
            self.logger.debug(
                f"Discarding msg by player {message.current_pose.player_id} "
                + f"in team {message.current_pose.team} at {message.timestamp.seconds}"
            )
            return

        team_data = self.protocol_converter.convert_from_message(message, self.create_team_data())
        self.team_data_publisher.publish(team_data)

    def send_message(self):
        if not self.is_robot_allowed_to_send_message():
            self.logger.debug("Robot is not allowed to send message")
            return

        now = self.get_current_time()
        msg = self.create_empty_message(now)

        def is_still_valid(time) -> bool:
            return now - Time.from_msg(time) < Duration(seconds=self.lifetime)

        message = self.protocol_converter.convert_to_message(self, msg, is_still_valid)
        self.socket_communication.send_message(message.SerializeToString())

    def create_empty_message(self, now: Time) -> Proto.Message:
        message = Proto.Message()
        seconds, nanoseconds = now.seconds_nanoseconds()
        message.timestamp.seconds = seconds
        message.timestamp.nanos = nanoseconds
        return message

    def create_team_data(self) -> TeamData:
        return TeamData(header=self.create_header_with_own_time())

    def create_header_with_own_time(self) -> Header:
        return Header(stamp=self.get_current_time().to_msg(), frame_id=self.map_frame)

    def should_message_be_discarded(self, message: Proto.Message) -> bool:
        player_id = message.current_pose.player_id
        team_id = message.current_pose.team

        is_own_message = player_id == self.player_id
        is_message_from_oposite_team = team_id != self.team_id

        return is_own_message or is_message_from_oposite_team

    def is_robot_allowed_to_send_message(self) -> bool:
        return self.gamestate is not None and not self.gamestate.penalized

    def get_current_time(self) -> Time:
        return self.node.get_clock().now()


def main():
    rclpy.init(args=None)
    TeamCommunication()


if __name__ == "__main__":
    main()
