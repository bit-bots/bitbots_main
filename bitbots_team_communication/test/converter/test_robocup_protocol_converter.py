from unittest.mock import MagicMock

import pytest
from soccer_vision_attribute_msgs.msg import Robot as RobotAttributes

import bitbots_team_communication.robocup_extension_pb2 as Proto  # noqa: N812
from bitbots_team_communication.converter.robocup_protocol_converter import RobocupProtocolConverter, TeamColor

own_team_color = TeamColor.BLUE


def test_setup_of_mappings(snapshot):
    converter = protocol_converter()
    assert converter.role_mapping == snapshot
    assert converter.action_mapping == snapshot
    assert converter.side_mapping == snapshot


def test_setup_of_team_color_mapping(snapshot):
    converter = protocol_converter()
    assert converter.state_to_proto_team_mapping[RobotAttributes.TEAM_OWN] == Proto.Team.BLUE
    assert converter.state_to_proto_team_mapping[RobotAttributes.TEAM_OPPONENT] == Proto.Team.RED
    assert converter.state_to_proto_team_mapping[RobotAttributes.TEAM_UNKNOWN] == Proto.Team.UNKNOWN_TEAM
    assert converter.proto_to_team_data_team_mapping == snapshot


def test_setup_of_to_message_converter(snapshot, to_message_mock):
    protocol_converter()

    to_message_mock.assert_called_once()
    args = to_message_mock.mock_calls[0].kwargs
    assert args == snapshot


def test_setup_of_from_message_converter(snapshot, from_message_mock):
    protocol_converter()

    from_message_mock.assert_called_once()
    args = from_message_mock.mock_calls[0].kwargs
    assert args == snapshot


def test_maps_convert_functions(to_message_mock, from_message_mock):
    to_message_mock.return_value.convert.return_value = "message"
    from_message_mock.return_value.convert.return_value = "team_data"

    converter = protocol_converter()

    assert converter.convert_to_message() == "message"
    assert converter.convert_from_message() == "team_data"


def protocol_converter() -> RobocupProtocolConverter:
    return RobocupProtocolConverter(own_team_color)


@pytest.fixture
def to_message_mock(mocker) -> MagicMock:
    return mocker.patch("bitbots_team_communication.converter.robocup_protocol_converter.StateToMessageConverter")


@pytest.fixture
def from_message_mock(mocker) -> MagicMock:
    return mocker.patch("bitbots_team_communication.converter.robocup_protocol_converter.MessageToTeamDataConverter")
