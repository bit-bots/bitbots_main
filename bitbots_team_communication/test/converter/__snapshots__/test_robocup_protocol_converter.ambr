# serializer version: 1
# name: test_setup_of_from_message_converter
  dict({
    'action_mapping': dict({
      0: 0,
      1: 1,
      2: 2,
      3: 3,
      4: 4,
      5: 5,
      6: 6,
      7: 7,
    }),
    'role_mapping': dict({
      0: 0,
      1: 1,
      2: 2,
      3: 3,
      4: 4,
      5: 5,
      6: 6,
    }),
    'side_mapping': dict({
      0: 0,
      1: 1,
      2: 2,
      3: 3,
    }),
    'team_mapping': dict({
      0: 1,
      1: 3,
      2: 2,
    }),
  })
# ---
# name: test_setup_of_mappings
  tuple(
    tuple(
      0,
      0,
    ),
    tuple(
      1,
      1,
    ),
    tuple(
      2,
      2,
    ),
    tuple(
      3,
      3,
    ),
    tuple(
      4,
      4,
    ),
    tuple(
      5,
      5,
    ),
    tuple(
      6,
      6,
    ),
  )
# ---
# name: test_setup_of_mappings.1
  tuple(
    tuple(
      0,
      0,
    ),
    tuple(
      1,
      1,
    ),
    tuple(
      2,
      2,
    ),
    tuple(
      3,
      3,
    ),
    tuple(
      4,
      4,
    ),
    tuple(
      5,
      5,
    ),
    tuple(
      6,
      6,
    ),
    tuple(
      7,
      7,
    ),
  )
# ---
# name: test_setup_of_mappings.2
  tuple(
    tuple(
      0,
      0,
    ),
    tuple(
      1,
      1,
    ),
    tuple(
      2,
      2,
    ),
    tuple(
      3,
      3,
    ),
  )
# ---
# name: test_setup_of_team_color_mapping
  dict({
    0: 1,
    1: 3,
    2: 2,
  })
# ---
# name: test_setup_of_to_message_converter
  dict({
    'action_mapping': dict({
      0: 0,
      1: 1,
      2: 2,
      3: 3,
      4: 4,
      5: 5,
      6: 6,
      7: 7,
    }),
    'role_mapping': dict({
      0: 0,
      1: 1,
      2: 2,
      3: 3,
      4: 4,
      5: 5,
      6: 6,
    }),
    'side_mapping': dict({
      0: 0,
      1: 1,
      2: 2,
      3: 3,
    }),
    'team_mapping': dict({
      0: 0,
      1: 1,
      2: 2,
    }),
  })
# ---
